### PR TITLE
feat(init): detect css modules and auto-install compatibility workarounds

### LINKED ISSUE #2992: CSS Modules: `@extend` (postcss-extend-rule) silently dropped — CSS-Modules plugin runs before user PostCSS plugins
## Problem

`@extend` from `postcss-extend-rule` is silently discarded in CSS Modules. Every declaration the rule should have inherited is missing — in dev and in the production build alike — with no warning or error.

```css
/* styles.module.css */
.shared { position: absolute; visibility: hidden; }
.wrap   { @extend .shared; border: 1px solid red; }
```

Expected: `.wrap` inherits `position: absolute; visibility: hidden`.

Actual output:

```css
._shared_1u9uh_1{visibility:hidden;position:absolute}
._wrap_1u9uh_2{border:1px solid red}
```

The at-rule is gone and nothing was copied.

## Cause

Vite's `compileCSS` puts the CSS-Modules plugin at the **front** of the PostCSS chain:

```js
if (isModule) postcssPlugins.unshift((await importPostcssModules()).default({ ... }))
```

So class names are already hashed to `._shared_1u9uh_1` by the time the project's `postcss-extend-rule` runs. `@extend .shared` matches no rule, and the plugin drops it silently.

Running the *same* PostCSS config standalone over the same file resolves the extend correctly, so the configuration is fine — only the ordering inside Vite's pipeline is wrong.

## Reproduction (no vinext involved)

```
postcss.config.mjs   ->  export default { plugins: { 'postcss-extend-rule': {} } }
src/styles.module.css -> the two rules above
src/main.js          ->  import s from "./styles.module.css"
vite build
```

The emitted CSS shows `_wrap_` without the extended declarations. This reproduces on plain Vite with no vinext in the project, so the underlying issue is Vite's plugin ordering — filing here because vinext already resolves and injects the project's PostCSS config, and is where a Next-compatible pipeline would be expected to handle it (webpack/`css-loader` runs the user's PostCSS before CSS-Modules scoping, so the same stylesheet works under `next build`).

## Impact

Any CSS module using `@extend` loses those declarations. In one Pages Router app this silently affected **46 `@extend` uses across 11 module files**, producing components with missing positioning — dropdown panels rendering in flow instead of as absolutely-positioned overlays, etc. Because nothing warns, it presents as "the CSS is just wrong" with no obvious cause.

## Workaround

Expanding `@extend` before Vite's pipeline sees the file, via an `enforce: "pre"` transform, fixes it — the extend resolves against the original selectors and Vite then hashes the already-expanded rules:

```ts
const expandCssModuleExtends = () => {
  let processor;
  return {
    name: "expand-css-module-extends",
    enforce: "pre" as const,
    async transform(code: string, id: string) {
      const file = id.split("?")[0];
      if (!file.endsWith(".module.css") || !code.includes("@extend")) return null;
      if (!processor) {
        const postcss = (await import("postcss")).default;
        const extendRule = (await import("postcss-extend-rule")).default;
        const presetEnv = (await import("postcss-preset-env")).default;
        // `@extend` must see flat selectors: postcss-extend-rule skips nested
        // rules whose selector it cannot resolve, so flatten nesting first.
        processor = postcss([
          presetEnv({ stage: false, features: { "nesting-rules": true } }),
          extendRule(),
        ]);
      }
      const result = await processor.process(code, { from: file });
      return { code: result.css, map: null };
    },
  };
};
```

Note the nesting step — without it only the `@extend`s at one nesting depth are expanded, which looks like a partial fix.

## Environment

- vinext `1.0.0-beta.6`
- Vite `8.1.0`
- `postcss-extend-rule`, `postcss-preset-env`
- Reproduced in dev and in the production build, and on plain Vite without vinext

### PR DESCRIPTION
**Pull request by NriotHrreion**
Closes #2992

## Overview

Vite scopes CSS Modules before running user PostCSS plugins, which differs from Next.js and can silently break transformations that depend on original selectors, such as `postcss-extend-rule`. `vite-css-modules` restores the expected processing order, but its default class-name hashing can produce different results across vinext's separate server and client environments (see https://github.com/cloudflare/vinext/issues/2992#issuecomment-5348417327).

This PR makes CSS Modules projects work consistently by automatically setting up `vite-css-modules` and deterministic, project-relative class names, preserving PostCSS behavior while preventing server/client class-name mismatches in development and production builds.

## What changed

- During `vinext init`, scan the original Next.js project folder for css modules, if there is, enable the css modules compatibility setup path.
- `vite-css-modules` will be installed as a devDependency.
- `vite-css-modules` will be added to vite config as a plugin, along with the `generateScopedName()` config.
- Edge cases of creating and modifying vite config file are considered and added to the tests.

## Verification

- Functionality of `vite-css-modules` is verified in the PR https://github.com/cloudflare/vinext/pull/3000
- `pnpm test tests/init.test.ts tests/init-cloudflare.test.ts`

## Question

Should we add an option (maybe a cli flag) to let developers decide whether to apply the css modules compatibility setup?

### CODE REVIEW

**Comment by james-elicx** [comment]
/bigbonk review for issues. 8 min time limit.

**Comment by james-elicx** [comment]
/bigbonk review for issues — time limit: 10 minutes; please return a concise clean verdict or actionable in-scope findings.

File: packages/vinext/src/init-cloudflare.ts [hunk 1]
```diff
@@ -1466,6 +1616,274 @@ function ensurePlugins(
...
+  const secondBindings = collectTopLevelBindings(secondProgram);
+  const secondOutput = new MagicString(withPlugin);
+  const preservedExistingGenerateScopedName = ensureCssModulesScopedName(
+    secondOutput,
+    secondConfig,
+    withPlugin,
+    (indent) => {
+      const existingCreateHash = commonJs
+        ? findRequiredBinding(secondProgram, "node:crypto", "createHash")
+        : findImportedBinding(secondProgram, "node:crypto", "createHash");
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts:2111]: Fixed in e691ce828: runtime import lookup now ignores declaration- and specifier-level type imports, and inserted value imports avoid `import type` declarations. Added regression coverage for both forms.

File: packages/vinext/src/init-cloudflare.ts [hunk 2]
```diff
@@ -1466,6 +1616,274 @@ function ensurePlugins(
...
+  preservedExistingGenerateScopedName: boolean;
+};
+
+/** Add the issue #2992 CSS Modules workaround without replacing user config. */
+export function updateViteConfigForCssModules(
+  filePath: string,
+  code: string,
+): CssModulesConfigUpdate {
+  const firstProgram = parseViteConfig(filePath, code);
+  const firstConfig = findConfigObject(firstProgram);
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts:2041]: Fixed in e691ce828 by reusing the existing transparent-expression unwrapping helper. Added coverage for `satisfies`, `as`, type assertions, and non-null assertions.

**Comment by james-elicx** [comment]
/bigbonk review for issues — time limit: 10 minutes; please return a concise clean verdict or actionable in-scope findings.

File: packages/vinext/src/init-cloudflare.ts [hunk 3]
```diff
@@ -570,27 +588,69 @@ function vinextExpression(
     : `${binding}({\n  ${optionEntries.join(",\n  ")},\n})`;
 }
 
+function generateScopedNameMethodSource(
+  indent: string,
+  pathBinding: string,
+  createHashBinding: string,
+  rootExpression = "import.meta.dirname",
+): string {
+  return `${indent}generateScopedName(name, filename) {
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts]: Fixed in 8951dd5c2: generated TS config methods now type both parameters; JS config variants keep the untyped JS-safe signature. Added coverage for both.

File: packages/vinext/src/init-cloudflare.ts [hunk 4]
```diff
@@ -78,6 +81,9 @@ export function validateCloudflarePlatformSetup(
         prerender: context.prerender,
       },
     );
+    if (context.hasCssModules) {
+      updateViteConfigForCssModules(context.existingViteConfigPath, cloudflareConfig);
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts]: Fixed in 8951dd5c2: `force` is threaded through Cloudflare preflight/setup and replaces unsupported dynamic css/modules values while preserving the rest of the config. Added an end-to-end Cloudflare force regression.

File: packages/vinext/src/init-cloudflare.ts [hunk 5]
```diff
@@ -1466,6 +1633,274 @@ function ensurePlugins(
...
+    return false;
+  }
+  if (modules.value.type !== "ObjectExpression") {
+    throw new Error(
+      "The Vite config's css.modules option must be a static object for vinext init to configure CSS Modules.",
+    );
+  }
+  const modulesObject = modules.value as AstObject;
+  const generateScopedName = findLastProperty(modulesObject, "generateScopedName");
+  if (generateScopedName) {
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts:2000]: Fixed in 8951dd5c2: null/undefined scoped-name values are replaced with the deterministic function and no longer reported as preserved. Added both regressions.

**Comment by james-elicx** [comment]
/bigbonk review for issues — time limit: 10 minutes; please return a concise clean verdict or actionable in-scope findings.

File: packages/vinext/src/init-cloudflare.ts [hunk 6]
```diff
@@ -1466,6 +1644,305 @@ function ensurePlugins(
...
+    if (hasPotentialSpreadOverride(modulesObject, generateScopedName)) {
+      throw new Error(
+        "The Vite config's css.modules.generateScopedName option must appear after any spread properties so vinext init can verify it.",
+      );
+    }
+    if (!hasNullishValue(generateScopedName)) return true;
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts]: Fixed in bc86aace0: string strategies containing Vite `[hash]` templates are replaced with the deterministic function; deterministic non-hash strings remain preserved. Added coverage.

File: packages/vinext/src/init.ts [hunk 7]
```diff
@@ -529,11 +591,18 @@ export async function init(options: InitOptions): Promise<InitResult> {
...
         isAppRouter: isApp,
         existingViteConfigPath,
+        force: options.force,
         prerender: options.prerender,
+        hasCssModules,
         today: options._today,
       },
       options.cloudflare!,
     );
+  } else if (hasCssModules && existingViteConfigPath && !options.force) {
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init.ts:602]: Fixed in bc86aace0: static objects now resolve through identifier-bound `defineConfig` variables and direct identifier arguments, with cycle protection. Both common shapes are covered.

File: packages/vinext/src/init.ts [hunk 8]
```diff
@@ -139,15 +140,51 @@ type InitResult = {
...
+/** Detect project-owned CSS, SCSS, or Sass module files. */
+export function scanCssModuleFiles(root: string): boolean {
+  try {
+    return fs
+      .globSync("**/*.module.{css,scss,sass}", {
+        cwd: root,
+        withFileTypes: true,
+        exclude: (entry) =>
+          entry.isDirectory() &&
+          (entry.name.startsWith(".") || CSS_MODULE_SCAN_IGNORES.has(entry.name)),
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init.ts]: Fixed in bc86aace0: dependency/cache names stay pruned anywhere, while dist/out/build/coverage are ignored only at the project root. Added a nested app/build source regression.

**Comment by james-elicx** [comment]
/bigbonk review for issues — time limit: 10 minutes; please return a concise clean verdict or actionable in-scope findings.

File: packages/vinext/src/init-cloudflare.ts [hunk 9]
```diff
@@ -1466,6 +1669,310 @@ function ensurePlugins(
...
+  const generateScopedName = findLastProperty(modulesObject, "generateScopedName");
+  if (generateScopedName) {
+    if (hasPotentialSpreadOverride(modulesObject, generateScopedName)) {
+      throw new Error(
+        "The Vite config's css.modules.generateScopedName option must appear after any spread properties so vinext init can verify it.",
+      );
+    }
+    const value = generateScopedName.value as AstNode & { value?: unknown };
+    const usesHashTemplate =
+      value.type === "Literal" &&
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts]: Fixed in 6cea508df: scoped-name values are unwrapped before inspection, and no-substitution template literals are treated as static strings. Literal, `as const`, and template regressions are covered.

File: packages/vinext/src/init-cloudflare.ts [hunk 10]
```diff
@@ -800,6 +929,9 @@ function findConfigObject(program: ESTree.Program): AstObject | undefined {
   if (firstArgument.type === "SpreadElement") return undefined;
   const argumentObject = unwrapObject(firstArgument);
   if (argumentObject) return argumentObject;
+  if (firstArgument.type === "Identifier") {
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts:1031]: Fixed in 6cea508df: CommonJS `module.exports = defineConfig(options)` now resolves the static identifier argument. Added focused coverage.

File: packages/vinext/src/init-cloudflare.ts [hunk 11]
```diff
@@ -751,7 +864,23 @@ function findVariableObject(program: ESTree.Program, name: string): AstObject |
...
         continue;
       }
-      return unwrapObject(declaration.init);
+      const direct = unwrapObject(declaration.init);
+      if (direct) return direct;
+      const initializer = unwrapExpression(declaration.init);
+      if (initializer?.type === "Identifier") {
+        return findVariableObject(program, initializer.name, seen);
+      }
+      if (initializer?.type !== "CallExpression" || initializer.arguments.length === 0) {
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts]: Fixed in 6cea508df: call traversal is restricted to recognized `defineConfig` identifiers/import aliases and namespace bindings; arbitrary factory calls are rejected. Added regressions for both cases.

**Comment by james-elicx** [comment]
/bigbonk review for issues — time limit: 10 minutes; please return a concise clean verdict or actionable in-scope findings.

File: packages/vinext/src/init-cloudflare.ts [hunk 12]
```diff
@@ -734,13 +815,74 @@ function findProperty(object: AstObject, name: string): AstProperty | undefined
...
+function unwrapObject(expression: ESTree.Expression): AstObject | undefined {
+  const unwrapped = unwrapExpression(expression);
+  return unwrapped?.type === "ObjectExpression" ? (unwrapped as AstObject) : undefined;
+}
+
+function isDefineConfigCall(program: ESTree.Program, call: ESTree.CallExpression): boolean {
+  const callee = unwrapExpression(call.callee);
+  if (callee?.type === "Identifier") {
+    return (
+      callee.name === "defineConfig" ||
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts]: Fixed in ac098776b: bare-name acceptance was removed; only defineConfig bindings imported/required from Vite (including aliases and namespaces) are traversed. A local-shadow regression now rejects the unsafe shape.

File: packages/vinext/src/init-cloudflare.ts [hunk 13]
```diff
@@ -1466,6 +1717,313 @@ function ensurePlugins(
...
+    }
+    const value = unwrapExpression(generateScopedName.value);
+    const staticValue =
+      value?.type === "Literal" && typeof value.value === "string"
+        ? value.value
+        : value?.type === "TemplateLiteral" && value.expressions.length === 0
+          ? (value.quasis[0]?.value.cooked ?? value.quasis[0]?.value.raw)
+          : undefined;
+    const usesHashTemplate = staticValue !== undefined && /\[hash(?::[^\]]*)?\]/i.test(staticValue);
+    if (!hasNullishValue(generateScopedName) && !usesHashTemplate) return true;
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts]: Fixed in ac098776b: nullish detection now receives the already-unwrapped value. Added `undefined as string | undefined` and `null as any` regressions.

**Comment by james-elicx** [comment]
/bigbonk review for issues — time limit: 10 minutes; please return a concise clean verdict or actionable in-scope findings.

File: packages/vinext/src/init-cloudflare.ts [hunk 14]
```diff
@@ -1466,6 +1718,313 @@ function ensurePlugins(
...
+}
+
+function ensurePluginFirst(
+  output: MagicString,
+  config: AstObject,
+  expression: string,
+  binding: string,
+  code: string,
+): void {
+  const plugins = findLastProperty(config, "plugins");
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts:1860]: Fixed in eea0a786c: static string and no-substitution template keys now resolve to their effective property names, while dynamic computed keys are rejected when they could override the configured property. Added regressions for computed plugins/css/modules and dynamic keys.

File: packages/vinext/src/init-cloudflare.ts [hunk 15]
```diff
@@ -1466,6 +1718,313 @@ function ensurePlugins(
...
+  const firstOutput = new MagicString(code);
+
+  const patchNamespace = commonJs
+    ? findRequiredBinding(firstProgram, "vite-css-modules", "default")
+    : findNamespaceImportedBinding(firstProgram, "vite-css-modules");
+  const existingMemberCall = findPluginMemberCall(firstConfig, patchNamespace, "patchCssModules");
+  const existingPatch = commonJs
+    ? findRequiredBinding(firstProgram, "vite-css-modules", "patchCssModules")
+    : findImportedBinding(firstProgram, "vite-css-modules", "patchCssModules");
+  if (!existingMemberCall) {
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts]: Fixed in eea0a786c: recognized namespace plugin calls now run the same later-spread/dynamic-key safety validation before being accepted. Added the reported namespace-plus-later-spread regression.

**Comment by james-elicx** [comment]
/bigbonk review for issues — time limit: 10 minutes; please return a concise clean verdict or actionable in-scope findings.

File: packages/vinext/src/init-cloudflare.ts [hunk 16]
```diff
@@ -1466,6 +1724,320 @@ function ensurePlugins(
...
+      )},\n${indent}  },\n${indent}},`,
+      code,
+      true,
+    );
+    return false;
+  }
+  if (css.value.type !== "ObjectExpression") {
+    if (force) {
+      const indent = objectPropertyIndent(config, code);
+      output.overwrite(
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts]: Fixed in 510fd2bc9: forced replacements now rewrite shorthand css/modules properties as explicit properties, preserving valid syntax. Added both shorthand regressions.

File: packages/vinext/src/init.ts [hunk 17]
```diff
@@ -230,11 +264,16 @@ export function getInitDependencyGroups(
     dependencies.push("@vinext/cloudflare");
     devDependencies.push("@cloudflare/vite-plugin", "wrangler");
   }
+  if (hasCssModules) devDependencies.push("vite-css-modules");
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init.ts]: Fixed in 510fd2bc9: CSS Modules projects now add postcss alongside vite-css-modules. Verified the published package contract requires postcss ^8.4.49 (with only lightningcss optional), and added dependency/install coverage.

File: packages/vinext/src/init-cloudflare.ts [hunk 18]
```diff
@@ -778,9 +947,18 @@ function findConfigObject(program: ESTree.Program): AstObject | undefined {
       }
       const direct = unwrapObject(expression.right);
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts:988]: Fixed in 510fd2bc9: direct CommonJS identifier exports now resolve through the same static object lookup used by ESM. Added the reported regression.

File: packages/vinext/src/init-cloudflare.ts [hunk 19]
```diff
@@ -1466,6 +1724,320 @@ function ensurePlugins(
...
+    : findImportedBinding(firstProgram, "vite-css-modules", "patchCssModules");
+  if (existingMemberCall) {
+    const plugins = findLastProperty(firstConfig, "plugins");
+    if (hasPotentialSpreadOverride(firstConfig, plugins)) {
+      throw new Error(
+        "The Vite config's plugins option must be explicitly defined after any spread or dynamic computed properties so vinext init can verify the existing CSS Modules plugin.",
+      );
+    }
+  } else {
+    const patchLocal = existingPatch ?? allocateBinding(bindings, "patchCssModules");
```
[the remark this answers is not in the document]
Comment (james-elicx) [reply packages/vinext/src/init-cloudflare.ts]: Fixed in 510fd2bc9: inserted import names are now allocated against bindings across the config, including callback parameters and locals. Added a callback collision regression covering patchCssModules, path, and createHash.

**Comment by james-elicx** [comment]
/bigbonk review for issues — time limit: 10 minutes; please return a concise clean verdict or actionable in-scope findings.

**Comment by james-elicx** [comment]
Addressed both timed Big Bonk findings in `ba59f326e`: CSS Module discovery now traverses project-owned hidden source directories while excluding explicit generated/cache directories, and static callback-body resolution now also works through variable-bound ESM and CommonJS `defineConfig` calls. Focused init suites pass (235/235), targeted format/lint/type checks pass, and `git diff --check` is clean.

**Comment by james-elicx** [comment]
/bigbonk review for issues — hard time limit: 10 minutes; review exact head ba59f326e59cca70b9c99552183bdde1bf8a0988 after resolving the hidden-source-directory scan and variable-bound ESM/CommonJS defineConfig callback findings. Be verdict-first and return explicit LGTM/no findings or only concise actionable in-scope findings before the deadline.

**Comment by james-elicx** [comment]
Addressed the callback-shadowing finding in `42fab6a58`. Existing ESM imports and CommonJS requires are now given stable top-level aliases when the resolved config callback shadows them; plugin, `createHash`, and `path` bindings are covered, including updater idempotency. Focused init suites pass (237/237), changed-file checks and `git diff --check` are clean.

**Comment by james-elicx** [comment]
/bigbonk review for issues — hard time limit: 10 minutes; final review of exact head 42fab6a58a6efe38865647029bb0645887499a46 after resolving callback-shadowed helper imports/requires. Be verdict-first: verify only hidden-source CSS Module discovery and static config callback resolution/binding safety for ESM and CommonJS, then return explicit LGTM/no findings or concise actionable findings before the deadline.

**Comment by james-elicx** [comment]
Addressed the CommonJS directive-prologue/TDZ finding in `a96c2841c`. Each generated alias is now inserted after the top-level declaration that initializes its reused binding. A directive-prefixed CommonJS callback regression covers all three helpers and repeat-update idempotency. Focused init suites remain 237/237 green; changed-file checks and `git diff --check` pass.

**Comment by james-elicx** [comment]
/bigbonk review for issues — hard time limit: 10 minutes; final exact-head review of a96c2841cfe5f6b5b926abc6dbff09f771440092 after placing CommonJS helper aliases after their source declarations. Verdict first: review only the cumulative hidden-source CSS Module scan and static ESM/CommonJS config callback resolution/binding safety, then post explicit LGTM/no findings or concise actionable findings before the deadline.

**Comment by james-elicx** [comment]
Addressed the hidden-filename finding in `1645f3bad`. CSS Module discovery now explicitly includes dot-prefixed filenames at the project root, below ordinary directories, and below project-owned hidden directories while retaining the known generated/cache exclusions. Focused init suites pass (239/239); changed-file checks and `git diff --check` are clean.

**Comment by james-elicx** [comment]
/bigbonk review for issues — hard time limit: 10 minutes; final exact-head review of 1645f3bad2efd0dbff54a8d3dad97a98d7a67c01 after adding root and nested dot-prefixed CSS Module filename discovery. Verdict first: inspect only the cumulative hidden-source scan and static ESM/CommonJS config callback resolution/binding safety, then post explicit LGTM/no findings or concise actionable findings before the deadline.
